### PR TITLE
Update to Android 9

### PR DESCRIPTION
**Pull request by kasnder**
Changes:
- Update of libraries
- Expose DNS cache
- Add custom DNS servers
- Add exlcusion of apps from VPN

### CODE REVIEW

File: antmonitorlib/src/main/java/edu/uci/calit2/antmonitor/lib/vpn/VpnClient.java [hunk 1]
```diff
@@ -366,7 +407,18 @@ public ParcelFileDescriptor buildTunInterface() {
...
         // Use Google's DNS server
-        builder.addDnsServer("8.8.8.8");
+        builder.addDnsServer(DNS_SERVER);
+
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
+            for (String excludedApp: excludedApps) {
+                try {
+                    builder.addDisallowedApplication(excludedApp);
+                } catch (PackageManager.NameNotFoundException ex) {
+                    //
```
Comment (nshuba) [inline antmonitorlib/src/main/java/edu/uci/calit2/antmonitor/lib/vpn/VpnClient.java:417]: Consider printing a warning here (perhaps at a debug level), to follow the rest of the AntMonitor style.

File: antmonitorlib/src/main/java/edu/uci/calit2/antmonitor/lib/vpn/VpnController.java [hunk 2]
```diff
@@ -163,14 +166,37 @@ public static void connectInBackground(Context context, IncPacketFilter incFilte
...
         VpnClient.prepareVPNconnection(incFilter, outFilter, incPacketConsumer, outPacketConsumer);
         Intent intent = new Intent(context, VpnClient.class);
-        // Specify that the servic should immediately establish the connection
+        // Specify that the service should immediately establish the connection
         intent.putExtra(VpnClient.EXTRA_CONNECT_ON_STARTUP, true);
-        context.startService(intent);
+
+        // TODO: Re-enabling on restart does not work by this method on Android 10. Why?
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            //ContextCompat.startForegroundService(context, intent);
```
Comment (nshuba) [inline antmonitorlib/src/main/java/edu/uci/calit2/antmonitor/lib/vpn/VpnController.java:176]: Helpful debug message? :)

**Review by nshuba — APPROVED** [review]
Thanks for this! I am so happy to see others using AntMonitor, making it better, and uncovering hidden features that we never exposed, such as the DNS cache :)

One comment is that originally we preferred to keep the ForwarderManager and TCPForwarder non-public so that if someone is using the library module as-is, the internal structures and complexities of the library module are hidden. I am approving this anyway, but it's just something to keep in mind in case you end up re-factoring and not needing those classes to be public.

**Comment by kasnder** [comment]
Thanks for clarifying and approving! I tried to keep the changes minimal because I wasn’t quite sure what your design choices were in the first place. :)

**Pull request merged**